File: ruby-gem/lib/models/localized_product.rb
# frozen_string_literal: true

# Class from DFC Business Ontology: #LocalizedProduct
require_relative '../core/semantic_object'
require_relative 'what_subject'


module DfcLinkmlConnector
  module Models
    class LocalizedProduct < WhatSubject
      SEMANTIC_TYPE = "dfc-b:LocalizedProduct".freeze

      # @return [String]
      attr_accessor :image

      # @return [Float]
      attr_accessor :cost

      # @return [Float]
      attr_accessor :quantity

      # @return [String]
      attr_accessor :constitued_by

      # @return [String]
      attr_accessor :consumed_by

      # @return [String]
      attr_accessor :reference

      # @return [String]
      attr_accessor :produced_by

      # @return [String]
      attr_accessor :represented_by

      # @param semanticId [String]
      # @param date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, image: nil, cost: nil, quantity: nil, constituedBy: nil, consumedBy: nil, reference: nil, producedBy: nil, representedBy: nil
      def initialize(semanticId, date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, image: nil, cost: nil, quantity: nil, constituedBy: nil, consumedBy: nil, reference: nil, producedBy: nil, representedBy: nil)
        super(semanticId, date: date, description: description, name: name, characteristicOf: characteristicOf, dimension: dimension)
        @image = image
        @cost = cost
        @quantity = quantity
        @constitued_by = constituedBy
        @consumed_by = consumedBy
        @reference = reference
        @produced_by = producedBy
        @represented_by = representedBy
        self.semanticType = "dfc-b:LocalizedProduct"
        registerSemanticProperty("dfc-b:LocalizedProduct:image", &method("image")).valueSetter = method("image=")
        registerSemanticProperty("dfc-b:LocalizedProduct:cost", &method("cost")).valueSetter = method("cost=")
        registerSemanticProperty("dfc-b:LocalizedProduct:quantity", &method("quantity")).valueSetter = method("quantity=")
        registerSemanticProperty("dfc-b:LocalizedProduct:constitued_by", &method("constitued_by")).valueSetter = method("constitued_by=")
        registerSemanticProperty("dfc-b:LocalizedProduct:consumed_by", &method("consumed_by")).valueSetter = method("consumed_by=")
        registerSemanticProperty("dfc-b:LocalizedProduct:has_reference", &method("reference")).valueSetter = method("reference=")
        registerSemanticProperty("dfc-b:LocalizedProduct:produced_by", &method("produced_by")).valueSetter = method("produced_by=")
        registerSemanticProperty("dfc-b:LocalizedProduct:represented_by", &method("represented_by")).valueSetter = method("represented_by=")
      end

      Core::SemanticObject.type_registry[SEMANTIC_TYPE] = self
    end
  end
end

File: typescript-connector/src/models/LocalizedProduct.ts
// Class from DFC Business Ontology: #LocalizedProduct
import { SemanticObject } from "../core/SemanticObject.js";
import { WhatSubject, type WhatSubjectParams } from "./WhatSubject.js";
import type { PhysicalProduct } from "./PhysicalProduct.js";
import type { QuantitativeValue } from "./QuantitativeValue.js";
import type { SuppliedProduct } from "./SuppliedProduct.js";

export interface LocalizedProductParams extends WhatSubjectParams {
  image?: string;
  cost?: number;
  quantity?: number;
  constituedBy?: string;
  consumedBy?: string;
  producedBy?: string;
  hasQuantity?: QuantitativeValue | string;
  hasReference?: SuppliedProduct | string;
  representedBy?: PhysicalProduct | string;
}
export class LocalizedProduct extends WhatSubject {
  static get SEMANTIC_TYPE(): string {
    return "dfc-b:LocalizedProduct";
  }

  image?: string;
  cost?: number;
  quantity?: number;
  constituedBy?: string;
  consumedBy?: string;
  producedBy?: string;
  hasQuantity?: QuantitativeValue | string;
  hasReference?: SuppliedProduct | string;
  representedBy?: PhysicalProduct | string;

  constructor(
    semanticId: string,
    params?: LocalizedProductParams,
  ) {
    super(semanticId, params);
    this.image = params?.image;
    this.cost = params?.cost;
    this.quantity = params?.quantity;
    this.constituedBy = params?.constituedBy;
    this.consumedBy = params?.consumedBy;
    this.producedBy = params?.producedBy;
    this.hasQuantity = params?.hasQuantity;
    this.hasReference = params?.hasReference;
    this.representedBy = params?.representedBy;
    this.semanticType = LocalizedProduct.SEMANTIC_TYPE;
    this.registerSemanticProperty("dfc-b:Image", () => this.image);
    this.registerSemanticProperty("dfc-b:cost", () => this.cost);
    this.registerSemanticProperty("dfc-b:quantity", () => this.quantity);
    this.registerSemanticProperty("dfc-b:constituedBy", () => this.constituedBy);
    this.registerSemanticProperty("dfc-b:consumedBy", () => this.consumedBy);
    this.registerSemanticProperty("dfc-b:producedBy", () => this.producedBy);
    this.registerSemanticProperty("dfc-b:hasQuantity", () => this.hasQuantity);
    this.registerSemanticProperty("dfc-b:hasReference", () => this.hasReference);
    this.registerSemanticProperty("dfc-b:representedBy", () => this.representedBy);
  }
  static {
    SemanticObject.typeRegistry.set(LocalizedProduct.SEMANTIC_TYPE, LocalizedProduct);
  }
}
